### ISSUE TITLE
[Question] Getting localStorage in 1.10.0

### ISSUE THREAD

**Issue by naknode** [microsoft/playwright#6258]
I'm trying to get the `localStorage` of an app via:

`const storage = await context.storageState();` and it outputs: `{ cookies: [], origins: [] }`

and then I try: 

```typescript
    const localStorage = await page.evaluate(() =>
      JSON.stringify(window.localStorage)
    );
    console.log(localStorage);
```

and I get: `page.evaluate: Evaluation failed: DOMException: Failed to read the 'localStorage' property from 'Window': Access is denied for this document.`

I know this page has `localStorage` because In between tests, I can see the `localStorage` in the Application tab.

**Comment by dgozman** [comment]
@naknode Most likely your page is at some restricted origin, e.g. `about:blank`. Try logging `page.url()` or `page.evaluate(() => window.location.href)` and see what it says.

We can look into this in detail, but we need a repro script to run locally.

**Comment by naknode** [comment]
I realized I wasn't doing a `page.goTo()` so I was getting that `about:blank`.  Once I fixed that...  I am able to get the `localStorage`

 The repo is pretty basic. It's a blank repo since we're having a separate repo for our e2e test.

So now that i do `const localStorage = await page.evaluate(() => window.localStorage);`, I get the `localStorage` I need.

Modifying it now is another matter. Doing `window.localStorage.setItem()` doesn't seem to work.

**Comment by dgozman** [comment]
> Modifying it now is another matter. Doing `window.localStorage.setItem()` doesn't seem to work.

I am not sure I can help with this, web apis are up to the browser. I'd suggest to open Developer Tools and play with local storage in the console. Take a look at our [Debugging Tools](https://playwright.dev/docs/debug) doc.

**Comment by dgozman** [comment]
@naknode It seems like you want to modify `localStorage` from your script. Could you please elaborate on the usecase? Meanwhile, using `evaluate` is what I would do, unless we introduce some kind of API for that. For the new API, we need to understand the usecases first though.

Just to confirm, after you use `evaluate` to set values in the local storage, does `context.storageState()` return you the values? This methods is primarily used for [authentication](https://playwright.dev/docs/auth/) purposes.

**Comment by naknode** [comment]
> @naknode It seems like you want to modify `localStorage` from your script. Could you please elaborate on the usecase? Meanwhile, using `evaluate` is what I would do, unless we introduce some kind of API for that. For the new API, we need to understand the usecases first though.
> 
> Just to confirm, after you use `evaluate` to set values in the local storage, does `context.storageState()` return you the values? This methods is primarily used for [authentication](https://playwright.dev/docs/auth/) purposes.

Sure. I have a localStorage that needs to change (permission-wise) and I'd like to modify it so I can test permissions on the page.

**Comment by sergioariveros** [comment]
Same here, we also need to inject some key/values in the localStorage, our case is because we are enabling multiple mocks from the localStorage to test specific scenarios, it would be nice if we can have something like `sessionStorage.setItem("key", "value");` to edit and enter new data

**Comment by mindplay-dk** [comment]
Well, `cookies()` has an `addCookies()` counterpart - it seems like `storageState()` ought to have an `addStorageState()` counterpart?

But it looks like the intended workflow is to use contexts - and inject cookies and storage at context creation time:

https://playwright.dev/docs/api/class-browser#browser-new-context-option-storage-state

The guide also explains how to "reuse authentication state", meaning cookies and storage:

https://playwright.dev/docs/auth/#reuse-authentication-state

(I don't like the example, which relies on disk storage - but looks like it will also accept a data structure like the one returned by `storageState()`, so you probably don't actually need to use a file here.)

(And support for `sessionStorage` is missing on both ends, but that's a separate issue.)

**Comment by o-lagatta** [comment]
Somehow Playwright during the execution is not getting the localStorage because is executing before the values are even set by the web application. So from my end after running the authentication, localStorage is empty, but if I check in Chromium the data is there. My understanding is that needs some delay before it reads the localStorage ? 

`(async () => {
  const browser = await chromium.launch({ headless: false });
  const page = await browser.newPage();

  
  await page.goto('http://localhost:4200');

  await page.fill('[name="loginfmt"]', 'email@email.com');
  await page.click('[type="submit"]');

  await page.fill('input[type="password"]', 'password');
  await page.click('[type="submit"]');

  await page.click('text=No');

  // (checking the application listing button - I know now I am logged In)
  await page.waitForSelector('.button.header__link');

  console.log('reading the local storage');
  // get the data from localstorage
  await page.evaluate(() => {
    const returner = {};
    **// Local storage is empty, seems the execution runs before the actual values are set in the page localStorage**
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      console.log('key', key);
      console.log('getItem', localStorage.getItem(key));
      returner[key] = JSON.stringify(localStorage.getItem(key));
    }
    return returner;
  });

  // tried again but still empty
  const localStorage = await page.evaluate(() =>
    JSON.stringify(window.localStorage)
  );

  console.log('localStorageVariable', localStorage);
})();
`

**Comment by hacknlove** [comment]
> For the new API, we need to understand the usecases first though.

Here an usecase:

https://plausible.io/docs/excluding-localstorage

If you do track with plausible, and you want to run tests on your production environment, you can set `localStorage.plausible_ignore=true`

`page.evaluate` seems good enough for this simple usecase

**Comment by dcerniglia** [comment]
I'm running up agains the same issues. Our use case is that we want to want to set baseUrl from a Jenkinsfile environment variable before the tests run. We're using a runtime library that reads local storage before loading and want to be able to pass different urls (dev, staging, prod, etc.) before the page loads. Essentially, I want to do something like this:
```
 beforeEach(() => {
  const testUrl = process.env.TEST_URL);
  if (testUrl) {
    const testConfig = {"baseUrl": testUrl, "theme": "test" };
    const testConfigJson = JSON.stringify(testConfig);
    window.localStorage.setItem('debug', testConfigJson);
  }
});
```

**Comment by dgozman** [comment]
@dcerniglia Look into [`page.addInitScript()`](https://playwright.dev/docs/api/class-page#page-add-init-script):

```js
beforeEach(async ({ page }) => {
  // ...
  await page.addInitScript(value => {
    window.localStorage.setItem('debug', value);
  }, testConfigJson);
});
```

**Comment by dcerniglia** [comment]
@dgozman That did it. Thank you!

**Comment by tetianaman** [comment]
Use case: when page opens, the default value of an input is token from a localStorage. 
Waiting for an API ;)

**Comment by mindplay-dk** [comment]
Does anyone have a workaround to set `localStorage` and `sessionStorage` for any `origin`?

You can't simply do this with `addInitScript` or `evaluate`, because the security `origin` won't match.

I've been trying to reference this code for a userland implementation of something similar:

https://github.com/microsoft/playwright/blob/ab9d5a0dc4744fbfad6be2ed219b81977b6aa97d/packages/playwright-core/src/server/browserContext.ts#L369-L389

Here's what I came up with:

```ts
async function injectStorageItem(context: BrowserContext, origin: string, type: "localStorage" | "sessionStorage", key: string, value: string) {
  const page = await context.newPage();

  initConsoleOutput(page);
  
  await page.route(
    () => true,
    route => { route.fulfill({ body: `<html></html>` }) }
  );

  await page.goto(origin, { waitUntil: "load" });

  await page.evaluate(
    ({ type, key, value }) => { window[type].setItem(key, value) },
    { type, key, value }
  );

  await page.close();
}
```

To see if this works, I do a `console.log(await context.storageState());` which shows `origins: []`, so it doesn't work.

I think my strategy is similar to that of `browserContext.setStorageState`: intercept all routes and just put an empty HTML document in there, navigate to the `origin` of the storage I'm trying to inject, then `evaluate` a function that calls e.g. `localStorage.setItem` to inject the state.

I've tried attaching `on("console")` and `on("pageerror")` handlers and put `console.log` statements all over the place - including the body of the function I pass to `evaluate`, where I was able to confirm that `origin === window.origin`, and so it seems like this *should* be working... it's on the right `origin`, and it doesn't error - but it just doesn't set anything in storage.

I suspect the magic missing piece is this [`internalCallMetadata`](https://github.com/microsoft/playwright/blob/ab9d5a0dc4744fbfad6be2ed219b81977b6aa97d/packages/playwright-core/src/server/instrumentation.ts#L95) thing, which is not supported by the public API?

Any idea what could be wrong?

Note that I've also tried injecting storage via CDP with [`DOMStorage.setDOMStorageItem`](https://chromedevtools.github.io/devtools-protocol/tot/DOMStorage/#method-setDOMStorageItem) and that didn't work either - it gives an error message saying "no frame matches the security origin" or something. I figured this was because that API doesn't actually allow injections of storage to a security origin that isn't currently open. There is barely any documentation for it, and I eventually gave up wiped the code in favor of the approach I described above. I wondering if I'm somehow running into the same security limitation with both approaches?

Any other ideas for a workaround?

**Comment by mindplay-dk** [comment]
Ha ha, it works, I messed up somewhere else.

Well, there you have a work around. 😉

**Comment by manuman94** [comment]
Thanks for that workaround @mindplay-dk !

**Comment by ippeiukai** [comment]
The `addInitScript` works, but still needs an improvement:
- the function registered to `addInitScript` runs on every navigation
  - we want to setup localStorage only once before the first navigation (we don't want it to be reset on next navigation)
- the function registered to `addInitScript` runs regardless of the origin (probably not a big deal in most cases)

I actually like the initScript approach, as it is not limited to localStorage (sessionStorage? no problem). Cookies need be set before the request indeed, but initScript feels enough for storages. Perhaps something similar to `addInitScript` method, but with an option to run only once for particular origin would do the job?

**Comment by IEvangelist** [comment]
> Ha ha, it works, I messed up somewhere else.
> 
> Well, there you have a work around. 😉

Where is the workaround? I need something that works in the .NET SDK, I'm testing an end-to-end scenario where a test user logs into my app. It then uses existing `localStorage` values to set the rendered culture is so that its appropriate translated version is displayed. I'm then going to validate that the page renders the correctly localized versions. This would be really useful, thank you ahead of time 🙏🏽.

**Comment by b5414** [comment]
for someone 

```js
await page.addInitScript(()=>{
	window.localStorage.setItem('key', 'value');
});
```

**Comment by Exoow** [comment]
Use case: without local storage, our application forces to user to fill in certain data before work can be continued. Setting local storage directly would speed up the tests.

**Comment by np-xavier** [comment]
Adding my own use case: Our applications have one-time modal confirmation boxes that appear when a new session is launched. Whether the modal shows or not is set and read in Local Storage, so being able to set them all to 'confirmed' at the start of the test would be much better than trying to wait to see if the modal is present and then make it go away with clicks.

**Comment by alcabon** [comment]
npm install node-localstorage

A drop-in substitute for the browser native localStorage API that runs on node.js.

https://www.npmjs.com/package/node-localstorage  ( Weekly Downloads: 183 185 )

const LocalStorage = require('node-localstorage').LocalStorage;
const localStorage = new LocalStorage('./scratch');  // local folder with the saved values

Same functions: localStorage.setItem localStorage.getItem,etc.

JSDOM instead of DOM and this drop-in substitute for localStorage. NodeJS runs on the server side.

**Comment by mykhailo-kobylianskyi** [comment]
> > for someone
> > ```js
> > await page.addInitScript(()=>{
> > 	window.localStorage.setItem('key', 'value');
> > });
> > ```
> 
> This should be put in the beforeAll/beforeEach block? Because I have done it that way and it is not consistent. Code: http://md-paster.herokuapp.com/63207fc3db3731001626e289

I've notices that it works well until as a parameter get string value directly like 'random'. However  nothing happen if I try to put here string variable. 
Will work:
``` window.localStorage.setItem('some-key', 'some-value')```

Will NOT work
``` const value : string  = "some-value"
 window.localStorage.setItem('some-key',  value)```

**Comment by Realman78** [comment]
It works fine if you're using the build to test and disabling the devTools. Dev tools probably have some kind of persistence with local storage which I couldn't get rid of. Make sure to DISABLE them, not just NOT OPEN.

**Comment by usersina** [comment]
By collecting the answers together and digging into the docs a bit, I've managed to implement my own `setStorageState` function which is the exact opposite of [**browserContext.storageState**](https://playwright.dev/docs/api/class-browsercontext#browser-context-storage-state)

Supposing you have the following folder structure:

<pre>
playwright-setup/
├── globalSetup.ts
├── helpers.ts
├── storageState.json // content can be initially set to `{"cookies": [],"origins": []}`
</pre>

`helpers.ts`
```typescript
import { Page } from 'playwright';

/**
 * Set the cookies as well as the localStorage items for each origin
 * to the context of the current page.
 *
 * This function causes an extra `page.goto` at the end. The `pageOptions`
 * param is for that navigation.
 *
 * @param page The page in question
 * @param param1 An object containing `cookies` and `origins`
 * @param pageOptions page options for the navigation
 */
export async function setStorageState(
  page: Page,
  { cookies, origins }: { cookies: Cookie[]; origins: Origin[] },
  pageOptions?: PageOptions,
) {
  const initialUrl = page.url();

  // 1. Set all origins which include the localStorage
  for (const { origin, localStorage } of origins) {
    // 1.1 Navigate to the origin
    await page.goto(origin);

    // 1.2 Loop through the items in localStorage and assign them
    for (const { name, value } of localStorage) {
      await page.evaluate(async (args: string) => {
        const [name, value] = args.split(',');
        window.localStorage.setItem(name, value);
        // console.log(JSON.stringify(window.localStorage));
      }, `${name},${value}`);
    }
  }

  // 2. Set cookies
  await page.context().addCookies(cookies.map((cookie) => ({ ...cookie, sameSite: <SameSite>cookie.sameSite })));

  // 3. Navigate back to the initial url
  await page.goto(initialUrl, pageOptions);
}

// =========================== Types =========================== //

interface PageOptions {
  referer?: string | undefined;
  timeout?: number | undefined;
  waitUntil?: 'load' | 'domcontentloaded' | 'networkidle' | 'commit' | undefined;
}

interface Cookie {
  name: string;
  value: string;
  domain: string;
  path: string;
  expires: number;
  httpOnly: boolean;
  secure: boolean;
  sameSite?: string;
}

enum SameSite {
  Lax = 'Lax',
  None = 'None',
  Strict = 'Strict',
}

interface Origin {
  origin: string;
  localStorage: LocalStorage[];
}

interface LocalStorage {
  name: string;
  value: string;
}

```

Finally, use it like this.

```typescript
import * as storageState from './storageState.json';

async function globalSetup(config: FullConfig) {
  ...
  await setStorageState(page, { cookies: storageState.cookies, origins: storageState.origins });
  ...
  // Save signed-in state back to 'storageState.json' just in case
  await page.context().storageState({ path: './playwright-setup/storageState.json' });
  await browser.close();
}

export default globalSetup;
```

<details> 
  <summary>For the sake of completeness, here's the global setup file if anyone is interested</summary>

`globalSetup.ts`
```typescript
async function globalSetup(config: FullConfig) {
  const browser = await chromium.launch({
    headless: !!process.env.CI,
  });
  const page = await browser.newPage();

  // Add login state from previous session if any
  await setStorageState(page, { cookies: storageState.cookies, origins: storageState.origins });

  // Go to the login page
  await page.goto('https://www.my-website.com/login/');

  // Check if we're already logged in by seeing if we got redirected to the dashboard
  const loggedIn = page.url().startsWith('https://www.my-website.com/dashboard/');

  // If not logged in, simply login manually
  if (!loggedIn) {
    await page.getByLabel('User Name').fill('username');
    await page.getByLabel('Password').fill('password');
    await page.getByText('Submit').click();
  }

  // Check if we finally logged in, if not throw an error
  if (!page.url().startsWith('https://www.my-website.com/dashboard/')) {
    await page.screenshot({ path: 'out/authentication.jpg', fullPage: true });
    throw new Error("Failed to login! See screenshot in 'out/authentication.jpg'");
  }

  // Save signed-in state to 'storageState.json' just in case something changed
  await page.context().storageState({ path: './playwright-setup/storageState.json' });
  await browser.close();
}

export default globalSetup;
```

<a href="https://thechive.com/wp-content/uploads/2022/06/9-619fa14d26075__700.jpg?attachment_cache_bust=4098028&quality=85&strip=info&w=400">One meme for the road</a>
</details>

**Comment by doublejosh** [comment]
I was hoping this thread would also reveal how to **read localStorage values** as part of an `expect` statement.

**Comment by pengwings** [comment]
> @dcerniglia Look into [`page.addInitScript()`](https://playwright.dev/docs/api/class-page#page-add-init-script):
> 
> ```js
> beforeEach(async ({ page }) => {
>   // ...
>   await page.addInitScript(value => {
>     window.localStorage.setItem('debug', value);
>   }, testConfigJson);
> });
> ```

Hi @dgozman,  I tried this and page.evaluate() but for some reason when I use page.evaluate(), my target value in local storage is set as "undefined" and when I use page.addInitScript(), my target value isn't changed at all.

```typescript
const authState = {
    users: {
      [authData.id]: authData.token,
    },
    currentUserId: authData.id,
    currentUser: null,
    currentWorkspaceId: null,
  };

  await page.evaluate(authState => {
    window.localStorage.setItem("AuthState", JSON.stringify(authState));
  });
```
I'm running this in a globalSetup function so I can share the login session across test cases.  `authData` comes from a response to a POST login request.  Is there something I'm missing?  Thanks in advance.

**Comment by giovanedann** [comment]
Hey guys, if someone face issues testing localStorage, this worked:

```
test('should save the access token on localStorage after submit succeeds', async ({
    page
  }) => {
    const accessToken = faker.datatype.uuid()

    await page.route('http://localhost:5050/api/login', async (route) => {
      const json = {
        accessToken
      }
      await route.fulfill({ status: 200, json })
    })

    await page
      .getByPlaceholder(/enter your e-mail/i)
      .fill(faker.internet.email())

    await page
      .getByPlaceholder(/enter your password/i)
      .fill(faker.internet.password(6))

    await page.getByRole('button', { name: /sign in/i }).click()

    expect(await page.evaluate(() => window.localStorage)).toStrictEqual({
      accessToken
    })
  })
```

**Comment by azamuddin** [comment]
@pengwings You need to reload the page with `page.reload()` to see localStorage in effect after calling `page.addInitScript`

> Hi @dgozman, I tried this and page.evaluate() but for some reason when I use page.evaluate(), my target value in local storage is set as "undefined" and when I use page.addInitScript(), my target value isn't changed at all.

**Comment by jeffersoneagley** [comment]
Use case: I have date selections for a filter set that are saved in `localStorage` and not being able to modify the local storage basically means the app can't reliably use playwright at all.

**Comment by WilliamSpademan** [comment]
Wow. This is a crucial issue for us. Our app uses localStorage a lot. The point of a behavioral test (as is baked into the Gherkin language) is to specify a starting state, suppose a specific action happens, and verify the result is as expected. Take the prototypical "Hello World!" app, for example, but add a button to ask for the user's name and store that in localStorage. Now when the app starts up we can have the main page say "Hello Jane!". We are disappointed Playwright cannot test such a basic feature. Our app stores a large volume of data when operating offline, for later upload to the server, so cookies are not a good option. We will use Selenium instead.

**Comment by GrayedFox** [comment]
I couldn't get this working by using `addInitScript()` or `evaluate()`, but passing in a `storageState` object when creating a new page did the trick.

To speed up our tests I wrote a helper library which allows developers to easily send requests directly against our GraphQL API and needed to copy information from some of the responses into local storage. Problem is the local storage must have the values set the moment our application embeds on a web page.

The only method that worked for me was passing in the `storageState` option when creating a new page, but I also wanted to avoid writing an extra test which writes to the disk and didn't want to use the CLI. I could not get [addInitScript to work as per the guide](https://playwright.dev/docs/auth/#reuse-authentication-state).

Here's a trimmed down version with everything placed together for brevity, including the types that PlayWright (at the time of writing) doesn't export. Could probably just use the MDN equivalent types, for now they are identical, but that could change.

```ts
// PlayWright type helpers
type CookieObject = {
  name: string;
  value: string;
  domain: string;
  path: string;
  expires: number;
  httpOnly: boolean;
  secure: boolean;
  sameSite: 'Strict' | 'Lax' | 'None';
};

type OriginObject = {
  origin: string;
  localStorage: { name: string; value: string }[];
};

type StorageState = {
  cookies: CookieObject[];
  origins: OriginObject[];
};

/**
 * A convenience method that should be called in each test's beforeAll hook that handles creating a page and browser instance.
 */
const setUp = async (
  browser: Browser,
  testInfo: TestInfo,
  testName: string,
  storageState?: StorageState
): Promise<[Page, Browser]> => {
  let page: Page;
  let testBrowser: Browser;
  // create page instance and set it as the testPage fixture
  if (isBrowserStack()) {
    [page, testBrowser] = await connectToBrowserStack(testInfo, testName);
  } else {
    testBrowser = browser;
    page = storageState
      ? await browser.newPage({ storageState })
      : await browser.newPage();
  }
  return [page, testBrowser];
};

test.describe.only('Payment GraphQL SDK', () => {
  let page: Page;
  let testBrowser: Browser;

  test.beforeAll(async ({ browser }, testInfo) => {
    // set up state, send requests to API, etc, before initialising page
    //  ... work work work ...
   
   const state: StorageState = {
      cookies: [],
      origins: [
        {
          origin: ClientData.baseUrl + '/index.html',
          localStorage: [{name: 'extreme', value: 'soup'}],
        },
      ],
    };

    // create page using state store
    [page, testBrowser] = await setUp(browser, testInfo, testInfo.title, state);

    // launch the browser
    await payment.open();
```

Took me a while to accept that I needed to create the page and launch the browser after hitting our API, since by habit the first thing nearly all of our tests do is create a page instance which is used to instantiate other page objects or connect to BrowserStack.

**Comment by jaredtmartin** [comment]
There have been several workarounds listed on this page, but I hope that this thread has shown that there is a real need for a simple command to set localStorage without having all these workarounds.

My use case: I want to set an authentication token to test as a logged in user without having to go through the entire login page process.

These workarounds are not adequate:
addInitScript:  I want to be able to test logging out, but this runs the script every time the user navigates to a different page, and since the site redirects the user to a certain page when they log out, it reruns the initScript, logging them back in as soon as they log out.
Global Setup:  I need some tests to be logged in and others not, so logging them in in global setup doesnt work.
new Context with storageState:  This seems very promising, but causes playwright to open browsers when running VSCode with the show browser option turned off, don't know how to prevent this. 

But in the end, these are just workarounds.  A proper setStorage function is definitely needed.  Even 
await page.evaluate(() => window.localStorage) seems like an overly complicated hack for such common functionality.

**Comment by nguyentramy87** [comment]
My use case: I want to place an order by bypassing the login screen with the token I get after an API call, which I want to be saved in `localStorage`.
I would love to see a setStorage function soon.

**Comment by Firdaus-prog** [comment]
> ```ts
> await page.evaluate(
>       `window.localStorage.setItem('X', 'Y')`
>     );
> ```

Can anyone confirm this?

**Comment by Firdaus-prog** [comment]
> Somehow Playwright during the execution is not getting the localStorage because is executing before the values are even set by the web application. So from my end after running the authentication, localStorage is empty, but if I check in Chromium the data is there. My understanding is that needs some delay before it reads the localStorage ?
> 
> `(async () => { const browser = await chromium.launch({ headless: false }); const page = await browser.newPage();
> 
> await page.goto('http://localhost:4200');
> 
> await page.fill('[name="loginfmt"]', '[email@email.com](mailto:email@email.com)'); await page.click('[type="submit"]');
> 
> await page.fill('input[type="password"]', 'password'); await page.click('[type="submit"]');
> 
> await page.click('text=No');
> 
> // (checking the application listing button - I know now I am logged In) await page.waitForSelector('.button.header__link');
> 
> console.log('reading the local storage'); // get the data from localstorage await page.evaluate(() => { const returner = {}; **// Local storage is empty, seems the execution runs before the actual values are set in the page localStorage** for (let i = 0; i < localStorage.length; i++) { const key = localStorage.key(i); console.log('key', key); console.log('getItem', localStorage.getItem(key)); returner[key] = JSON.stringify(localStorage.getItem(key)); } return returner; });
> 
> // tried again but still empty const localStorage = await page.evaluate(() => JSON.stringify(window.localStorage) );
> 
> console.log('localStorageVariable', localStorage); })(); `

can anyone confirm this?

**Comment by seanlaff** [comment]
I believe you can control local storage by creating your own browser->page rather than relying on the default. I do this to make a window the opts out of my default storage state:
```js
    const browser2 = await browser.newContext({ storageState: undefined });
    const page2 = await browser2.newPage();
```

**Comment by AlexWebDev01** [comment]
> My use case: I want to place an order by bypassing the login screen with the token I get after an API call, which I want to be saved in `localStorage`. I would love to see a setStorage function soon.

That's a common use case. Solving the same problem I found out that Feature request, so I'll share with the root cause and the solution of such problem.

**Use case**
Developing automation for one of our products, I get the AuthToken from the response when I register the user. I want to immediately set up that token to the localStorage of my page, so I can be already authenticated opening another pages.

**Implementation**
I initialise frontendPage as a fixture that uses prepared frontendContext fixture:
`
    frontendPage: async ({ frontendContext }, use) => {
        const frontendPage = await frontendContext.newPage();
        
        await use(frontendPage);
    },
`

Then I create a dedicated method to login after getting authToken during registration:
`
    async login(token: string) {
        const authValue = JSON.stringify({'token': `\"${token}\"`,'_persist':'{"version":-1,"rehydrated":true}'}) // Redux persist format
        await this.frontendPage.evaluate((token) => localStorage.setItem('persist:auth', token), authKey);
        console.log('Check if token is set up', await this.frontendPage.evaluate(() => localStorage.getItem('persist:auth')));
        
    }
`
![Screenshot 2024-03-28 at 17 19 04](https://github.com/microsoft/playwright/assets/102874212/ba46b872-987d-4bd8-a167-f1bcee40fd5d)

**Error**
Using that method I get such error:
`Error: page.evaluate: SecurityError: The operation is insecure.` pointing on .evaluate() method.
<img width="987" alt="Screenshot 2024-03-28 at 15 57 32" src="https://github.com/microsoft/playwright/assets/102874212/62290f80-e7be-464f-bdc6-9e9a722a8429">

**Solution**
The main point of that problem is:
If we have .addCookies() method, provided by Playwright, we also want to use some .setLocalStorageItem() method.
But the problem is we can't set something to the localStorage for blank/not existing page.
So all attempts to evaluate the blank/not existing page and trying to set the localStorage item face some errors.
To solve that problem you need to open any page with .goto() method. If you defined baseUrl ([check docs](https://playwright.dev/docs/test-webserver#adding-a-baseurl)) you can just open baseUrl by:
`
    await page.goto(pageUrl);
`
and only then try to set the localStorage item, evaluating the page:
`
       await page.evaluate((token) => localStorage.setItem('key', token), authKey);        
`

In my case the solution was:
`
    async login(token: string) {
        const authValue = JSON.stringify({'token': `\"${token}\"`,'_persist':'{"version":-1,"rehydrated":true}'}) // Redux persist format
        await this.frontendPage.goto('');
        await this.frontendPage.evaluate((token) => localStorage.setItem('persist:auth', token), authValue);        
    }
`
It's a mandatory to re-check what url uses your .goto() method. It takes baseUrl and add a string that you pass in it. Also check [docs](https://playwright.dev/docs/api/class-page#page-goto) to understand how it works.

And don't forget that you need to reload the page or go to another one after setting something to localStorage to make browser apply it.

So the step-by-step plan to set something to localStorage is:
1. Open existing & not blank page by using `goto()` method
`await this.frontendPage.goto('');`
2. Evaluate the page and use browser API `localStorage.setItem()`
`await this.frontendPage.evaluate((token) => localStorage.setItem('persist:auth', token), authValue);        
`
**Conclusions**
So I think all requests to create a built-in Playwright method to set item to the localStorage are not valid, because it's not the method that works with Playwright BrowserContext, but that's a browser API that works directly with opened web-page.

Hope my thoughts would be useful for community!

**Comment by rudreshtrivedi** [comment]
What I did was, addInitScript

`    await page.context().addInitScript(() => {
        localStorage.setItem('xxx', 'yyyy);
    });`

This solved the problem

**Comment by BenjaminLesne** [comment]
## Add a specific local storage to all tests:

```jsonc 
// playwright.config.ts
{
  // ...
  use: {
    storageState: {
      origins: [
        {
          localStorage: [
            { name: "myKey", value: "true" },
          ],
          origin: "https://example.com",
        },
      ],
      cookies: [],
    },
  },
  // ...
}
```

### What if some tests shouldn't have this local storage?

add the following in the wanted test files:
```ts
// myTest.spec.ts

test.describe("my beautiful tests", () => {
  test.use({
    storageState:  {
      origins: [
        {
          localStorage: [],
          origin: "https://example.com",
        },
      ],
      cookies: [],
    },
  });
    // ...
});
```

## sources:
- https://playwright.dev/docs/api/class-test#test-use
- https://playwright.dev/docs/api/class-testoptions#test-options-storage-state

At first I found this doc that made me struggle more than anything:
- https://playwright.dev/docs/auth#session-storage

**Comment by cbarrerah** [comment]
Our use case is that we are testing a web page that has some components that define areas of the app, and by default we use some already published packages, but each component repository runs some tests for their app, so we need to override the default available app with the one that we are locally serving before the execution of the tests.

We can do this by modifying the localStorage to include some entries.
It is true that it can be done with the page.eval() but it is quite cumbersome as it involves a lot of manipulation, and it introduces some confussion, as you can obtain the localstorage with`await page.context().storageState()` but there is no way to "save" any modifications to that said storageState to the browser instance.

In our case it is important, because we also need to login, and we save the storageState after login and reuse it between tests to save time on logins for different spec files. The thing is that if we don't use the `page.eval()` strategy, the storageState saved after the authentication, will not contain the app registration.

We solved it by doing the app registration with page.eval for the set item, but again, it is weird that we don't have a direct access to the saving of the localStorage to the browserContext as we have to the cookies.

**Comment by akshaypilankar** [comment]
Here is what worked for me


 // Navigate to the page and wait for it to load
  await page.goto('https://example.com', { waitUntil: 'domcontentloaded' });

  // Set an item in localStorage
  await page.evaluate(() => {
    localStorage.setItem('key', 'value');
  });
